src/core/Akka.Remote/EndpointManager.cs

```diff
@@ -492,196 +543,248 @@ protected void Accepting(object message)
+                if (policy.Item1 is Pass && policy.Item2 == null)
+                {
+                    var endpoint = policy.Item1.AsInstanceOf<Pass>().Endpoint;
+                    Context.Stop(endpoint);
+                    _log.Warning("Association to [{0}] with unknown UID is reported as quarantined, but " +
+                    "address cannot be quarantined without knowing the UID, gating instead for {1} ms.", quarantine.RemoteAddress, _settings.RetryGateClosedFor.TotalMilliseconds);
+                    _endpoints.MarkAsFailed(endpoint, Deadline.Now + _settings.RetryGateClosedFor);
+                }
+                else if (policy.Item1 is Pass && policy.Item2 != null)
+                {
+                    var pass = policy.Item1 as Pass;
+                    if (pass.Uid == quarantine.Uid)
                         Context.Stop(pass.Endpoint);
-                        if (!pass.Uid.HasValue)
-                        {
-                            log.Warning("Association to [{0}] with unknown UID is reported as quarantined, but address cannot be quarantined without knowing the UID, gated instead for {0} ms",
-                                quarantine.RemoteAddress, settings.RetryGateClosedFor.TotalMilliseconds);
-                            endpoints.MarkAsFailed(pass.Endpoint, Deadline.Now + settings.RetryGateClosedFor);
-                        }
+                }
+                else
+                {
+                        // Do nothing, because either:
+                        // A: we don't know yet the UID of the writer, it will be checked against current quarantine state later
+                        // B: we know the UID, but it does not match with the UID to be quarantined
                     }
 
-                    //Stop inbound read-only association
-                    var read = endpoints.ReadOnlyEndpointFor(quarantine.RemoteAddress);
-                    if (read != null)
+                    // Stop inbound read-only associations
+                    var readPolicy = Tuple.Create(_endpoints.ReadOnlyEndpointFor(quarantine.RemoteAddress), quarantine.Uid);
+                if (readPolicy.Item1?.Item1 != null && quarantine.Uid == null)
+                    Context.Stop(readPolicy.Item1.Item1);
+                else if (readPolicy.Item1?.Item1 != null && quarantine.Uid != null && readPolicy.Item1?.Item2 == quarantine.Uid) { Context.Stop(readPolicy.Item1.Item1); }
+                else { } // nothing to stop
+
+                    Func<AkkaProtocolHandle, bool> matchesQuarantine = handle => handle.RemoteAddress.Equals(quarantine.RemoteAddress) &&
+                                                                             quarantine.Uid == handle.HandshakeInfo.Uid;
+
+                    // Stop all matching pending read handoffs
+                    _pendingReadHandoffs = _pendingReadHandoffs.Where(x =>
+                {
+                    var drop = matchesQuarantine(x.Value);
```

Comment:
if any of the endpoints we are pending a read-handoff for match the quarantine address and UID, drop them as we iterate through the colletion